VERSION

```diff
@@ -0,0 +1 @@
+2.3.0
```

Comment:
If we needed this file, we would probably still add it to `.gitignore`?

Reply:
Let's discuss in this thread also the other questions surrounding this file.

> Does this need to be a file instead of a value given as a parameter?

It has to be a file, because nix flakes are [evaluated hermetically](https://www.tweag.io/blog/2020-07-31-nixos-flakes/#hermetic-evaluation), so we can't pass the version a parameter. We can't add the file to `.gitignore` either, because files used in a flake builds must be part of the working tree. Some alternatives I thought of:

1. Use the flake only to provide a shell with dependencies in which the binaries can be built, then it doesn't need to know about the version. I tried this initially, but ran into problems with some dependencies not being found.
2. We could make the flake be the source of truth in regards to the version, and expose it as an environment variable in the devshell. We could even include the git revision in the version string (almost like what `git describe` gives us):
```nix
      rev = if self ? shortRev then self.shortRev else "unknown";
      version = "2.3.0-${rev}";
```
This is only possible if the git tree is clean, which I guess is a good thing. Building with uncommitted changes, and using the last revision would not really be honest.

I'm wondering however if we even need the revision for the versions we release? As I understand, the binaries that are cross-built with nix are meant for deploying new releases. Wouldn't we make new releases from specific tags? In this case adding the revision seems redundant to me. Or do we also want to deploy releases from non-tagged commits?

Reply:
Agree to move the version's source of truth to the flake file. I find it awkward to duplicate the version in the source and the tag, but we do that elsewhere and it works.

I think crossbuilding is not a release-specific activity. When I want to test something I may compile on my Linux machine and test on a Windows machine. I may not want to bother to set up the Windows machine to compile there. I would rather not have to commit each time I trigger a build.

There seems to be `dirtyRev` and `dirtyShortRev` too[^1][^2].

[^1]: https://discourse.nixos.org/t/flakes-accessing-selfs-revision/11237/7
[^2]: https://github.com/NixOS/nix/pull/5385

Reply:
Oh, but we don't want `"2.3.0-${rev}"` when the rev is tagged, then we just want `"2.3.0"`.

I am not sure this is possible with flakes, because they make a point of throwing away useful information. :upside_down_face: 

How about

- making `flake.nix` the source of truth for the version,
- (while duplicating the version in a Git tag),
- only put a human-picked version string in the linked variable,
- include the revision as a separate variable and put into metadata [here](https://github.com/dividat/driver/blob/2aaf5787864a199a928404f3bea8cc659de52022/src/dividat-driver/server/main.go#L76-L82)

?

Reply:
I see what you mean, I agree that keeping the `rev` as part of the version string is useful. Your suggestion sounded good but as I started working on it, I realized that it might not work so well compared to `git describe`. 
Checking out some tag, `git describe` give us just the tag, without anything extra:

```
❯ git checkout 2.3.0
Note: switching to '2.3.0'. 
❯ git describe
2.3.0
```

We can't replicate this with the information provided by nix unfortunately. I find the `<version>-<hash>` option which we can do with nix much less useful, because it doesn't tell us where we are in relation to the closest tag. Building from the tag would look pretty much the same as building from any later commit, we'd have to look at `git log` to make sense of it. We can't have 

I had another go at providing the cross-compilation toolchains through shells instead of building actual derivations and this time I got it to work. I think we can go with that and keep using `git describe` for determining the version.

Reply:
[Provide cross-building toolchain in nix shells](https://github.com/dividat/driver/pull/129/commits/f8907ae140d171ca4444585e862ff1557fe1a9be)